### DISCUSSION TITLE
Define object methods

### DISCUSSION THREAD

**Discussion by hydroper** [grain-lang/grain#1882] [Ideas]
Grain looks very cool, but I've searched a little about object methods and found nothing. It looks like instead of `"".charAt(i)` you've to do `String.charAt(i, "")`. It'd be nice if there was something like:

```
let charAt = (i, self) => ...
```

Then you can do `"".charAt(i)`. You should be able to put `self` anywhere.

**Comment by alex-snezhko** [comment]
Grain doesn't leverage the object oriented paradigm very heavily so being able to attach methods to types likely won't be added, but when #403 and #402 get implemented (they're on the radar but aren't an immediate priority right now) you'd be able to use `|>` as an approximation for `.` when doing `object.method`. So for example you could represent what you'd write as `list.map(x => x * 2).reverse()` in an object-oriented language as `list |> partial List.map(x => x * 2, _) |> List.reverse` in Grain.

**Reply by hydroper** [reply]
I never understood this `|>` in functional languages: so it's basically a way to bind a function to some arguments, right? Interesting

**Reply by alex-snezhko** [reply]
Kind of, I believe the `|>` syntax was popularized by F# and it just applies the left side operand as a single argument to the right side function. Since F# uses curried function application, if it used Grain-style syntax it would just be `let (|>) = (arg, fn) => fn(arg)`. Other non-curried languages like Elixir and Gleam borrowed this syntax as well to mean "apply the LHS as the _first_ argument to the RHS function"
